### PR TITLE
[New features][Bug fixes] Add TileLang CSA compressed indexer for DSv4 and fix ColumnLinear stop_gradient backward

### PR DESCRIPTION
**Pull request by YJMSTR**
### PR Category
Operator Mechanism

### PR Types
New features, Bug fixes

### Description
1. **Bug fix（`tensor_parallel/layers.py`）**：修复 `ColumnLinearWithGradAccum` PyLayer backward 的合约问题。当 forward 输入 `input.stop_gradient=True` 时，backward 对应位置必须返回 `None`（Paddle PyLayer 合约要求）。在 forward 缓存 `input_stop_gradient` 标志，在 backward 中跳过 `grad_input` 计算及输入侧通信（all_reduce / reduce_scatter）。

2. **新功能（`tilelang_ops/indexer/`）**：新增 DeepSeek V4 CSA 压缩索引器 TileLang 融合算子：
   - `csa_indexer_topk_fwd`：流式 top-k 前向，输出索引和 softmax 概率
   - `csa_indexer_bwd`：融合反向，计算 IndexQ / Weights / IndexKComp 梯度
   - `csa_attn_target_reducesum`：indexer KL loss 的 attention target 分布
   - 新增 `TileLangCSAIndexerLoss` PyLayer 将 fwd/bwd/loss 融合为单一算子

3. **集成（`transformer/csa_attention.py`）**：通过 `dsv4_tilelang_backend` / `dsv4_tilelang_enable_csa_indexer` / `dsv4_tilelang_enable_backward` 三个 TransformerConfig 字段控制 TileLang 路径启用。默认关闭，不影响现有行为。

4. **简化**：移除 `attn_mask_startend_row_indices` 多文档边界支持（目前 DSv4 场景下不使用）。

### CODE REVIEW

File: src/paddlefleet/transformer/csa_attention.py [hunk 1]
```diff
@@ -372,6 +264,261 @@ def unfused_compressed_sparse_attn(
...
+            target,
+        ) = ctx.saved_tensor()
+
+        # Match the forward KL smoothing exactly:
+        # loss = target * (log(target + eps) - log(prob + eps)).
+        # First differentiate w.r.t. probabilities, then apply the softmax
+        # Jacobian for the selected logits.
+        eps = 1e-10
+        scale = ctx.loss_coeff / max(ctx.num_rows, 1.0)
+        grad_probs = -target / (topk_probs + eps) * scale
```
Comment (Enigmatisms) [inline src/paddlefleet/transformer/csa_attention.py]: 这地方能用 softmax + KL gradient 的快速公式吗？看着很像。因为是对 logits(假设是z) 求导，z --> softmax --> KL 之后 dL/dz 好像就直接是（topk_probs - target）（貌似要乘以 scale）?

File: src/paddlefleet/transformer/csa_attention.py [hunk 2]
```diff
@@ -887,19 +1114,42 @@ def forward(
...
+                    with paddle.no_grad():
+                        q_indexer_tl, k_indexer_tl, weights_indexer_tl = (
+                            self.indexer.forward_before_topk(x_det, qr_det)
+                        )
+                        tl_topk_indices, _tl_topk_scores = csa_indexer_topk_fwd(
+                            q_indexer_tl,
+                            k_indexer_tl,
+                            weights_indexer_tl,
+                            ratio=self.compress_ratio,
+                            topk_effective=topk_effective,
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/csa_attention.py]: P1 **优先级：P1**

这里在 `dsv4_tilelang_backend='attention_paddle_compat'` 且 `dsa_indexer_use_sparse_loss=False` 时会把 `topk_effective` 设成 `n_compressed`，随后这组全量候选会直接替换 `topk_indices_compressed` 并进入 sparse attention。默认 Paddle 路径虽然用全量候选计算 indexer loss，但实际 attention 仍只使用 `FusedDSAIndexerLoss._last_topk_indices`，其长度是 `min(index_topk, n_compressed)`；因此只打开 TileLang 前向兼容后端就会把每个 query 的压缩 KV attention fan-in 从 `index_topk` 扩大到 `n_compressed`，改变前向语义并显著增加注意力开销。建议 loss 用的 `topk_effective` 和 attention 用的 top-k 分开：loss/backward 可保留全量候选，前向替换 sparse attention indices 时仍按 `min(self.indexer.index_topk, n_compressed)` 生成或截取。

参考修复形态：

```python
attention_topk = min(self.indexer.index_topk, n_compressed)
loss_topk_effective = _resolve_csa_indexer_topk_effective(
    self.config,
    self.indexer.index_topk,
    n_compressed,
)

# TileLang loss path uses loss_topk_effective.
# Forward-only TileLang override uses attention_topk.
tl_topk_indices, _ = csa_indexer_topk_fwd(
    q_indexer_tl,
    k_indexer_tl,
    weights_indexer_tl,
    ratio=self.compress_ratio,
    topk_effective=attention_topk,
)
```

File: src/paddlefleet/transformer/dsv4_hybrid_attention.py [hunk 3]
```diff
@@ -225,9 +225,6 @@ def forward(
             attention_mask,
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/dsv4_hybrid_attention.py:227]: P1 **优先级：P1**

这里删除了 `attn_mask_startend_row_indices` 的下传，但 `DSv4HybridSelfAttention.forward()` 仍然通过 `**kwargs` 静默接收这个参数；上层 `TransformerLayer` 在 packed sequence / 多文档场景会继续把该边界 mask 传给 self-attention。结果是调用方以为文档边界仍然生效，CSA 实际却按整段序列生成 window/compressed indices，可能跨文档互相 attend。若本 PR 确认不再支持 DSv4 packed/multi-document 边界，建议在这里显式 fail fast；否则需要恢复边界 mask 到 `CompressedSparseAttention` 的索引计算。

参考修复形态：

```python
if kwargs.get("attn_mask_startend_row_indices") is not None:
    raise NotImplementedError(
        "DSv4 Hybrid CSA no longer supports attn_mask_startend_row_indices; "
        "please disable packed/multi-document boundary masks or restore CSA boundary handling."
    )
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已完成初轮审查，发现两处需要先修复的问题，细节已放在行级评论中。CI 当前通过，但这两处会分别改变 TileLang 兼容后端的 sparse attention 前向语义，以及让 packed/multi-document 边界 mask 被静默忽略，建议修复后再合入。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: src/paddlefleet/tilelang_ops/indexer/csa_indexer.py [hunk 4]
```diff
@@ -0,0 +1,418 @@
...
+def _contiguous(tensor):
+    return tensor.contiguous()
+
+
+def _cast_int32(tensor):
+    return tensor.cast("int32") if tensor.dtype != paddle.int32 else tensor
+
+
+def _cast_float32(tensor):
+    return tensor.cast("float32") if tensor.dtype != paddle.float32 else tensor
```
Comment (Enigmatisms) [inline src/paddlefleet/tilelang_ops/indexer/csa_indexer.py]: 是不是多余的？

File: src/paddlefleet/tilelang_ops/indexer/csa_indexer.py [hunk 5]
```diff
@@ -0,0 +1,418 @@
...
+    _DIGEST_COUNTERS[op] = count + 1
+    payload = {
+        "op": op,
+        "call": count,
+        "tensors": [
+            _tensor_digest(name, tensor) for name, tensor in tensors.items()
+        ],
+    }
+    print("[TileLangDigest] " + json.dumps(payload, sort_keys=True), flush=True)
+
```
Comment (Enigmatisms) [inline src/paddlefleet/tilelang_ops/indexer/csa_indexer.py]: 多余，之前有做过删除操作但是没同步上。

File: src/paddlefleet/tilelang_ops/indexer/csa_indexer.py [hunk 6]
```diff
@@ -0,0 +1,418 @@
...
+            f"index_k_comp must have shape [B, S_comp, D_i], got {tuple(index_k_comp.shape)}"
+        )
+    if len(weights.shape) != 3:
+        raise ValueError(
+            f"weights must have shape [B, S, H_i], got {tuple(weights.shape)}"
+        )
+
+    batch, seq_len, heads, dim = tuple(index_q.shape)
+    batch_k, _, dim_k = tuple(index_k_comp.shape)
+    batch_w, seq_len_w, heads_w = tuple(weights.shape)
```
Comment (Enigmatisms) [inline src/paddlefleet/tilelang_ops/indexer/csa_indexer.py]: tuple 包一下是为了保护什么？

File: src/paddlefleet/tilelang_ops/indexer/csa_indexer_bwd.py [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+    batch_k, seq_len_comp, dim_k = index_k_comp.shape
+    batch_i, seq_len_i, topk_effective = topk_indices.shape
+    batch_g, seq_len_g, topk_g = grad_scores.shape
+
+    assert batch == batch_w == batch_k == batch_i == batch_g
+    assert seq_len == seq_len_w == seq_len_i == seq_len_g
+    assert heads == heads_w
+    assert dim == dim_k
+    assert topk_effective == topk_g
+    assert topk_effective > 0
```
Comment (Enigmatisms) [inline src/paddlefleet/tilelang_ops/indexer/csa_indexer_bwd.py]: 挺多 assert 无描述的。

Comment (Enigmatisms) [reply src/paddlefleet/tilelang_ops/indexer/csa_indexer_bwd.py]: 不过不是什么大问题。

Comment (Enigmatisms) [reply src/paddlefleet/tilelang_ops/indexer/csa_indexer_bwd.py]: 不过说实在的，输入校验用 assert 确实容易被跳过（-O）。

**Review by risemeup1111 — COMMENTED** [review]
已复查当前 head。之前两条 P1 行级问题在当前 diff 中仍由原线程覆盖，本次不重复开新线程；CI 当前通过。另有一个非行级 PR 描述问题需要同步更新：

- P3 **优先级：P3** 非行级：PR 描述与当前配置字段不一致。描述第 3 点仍写 `dsv4_tilelang_backend` / `dsv4_tilelang_enable_csa_indexer` / `dsv4_tilelang_enable_backward`，但当前代码已改为 `csa_tilelang_backend` / `csa_tilelang_enable_indexer` / `csa_tilelang_enable_sparse_attn`，并且不再保留 `dsv4_tilelang_enable_backward`。这会误导使用者按无效字段配置 TileLang 路径。建议把该段改为：`通过 csa_tilelang_backend / csa_tilelang_enable_indexer / csa_tilelang_enable_sparse_attn 三个 TransformerConfig 字段控制 TileLang 路径启用。默认关闭，不影响现有行为。`

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Review by From00 — APPROVED** [review]

**Review by sneaxiy — APPROVED** [review]
